Sources/ServiceDiscovery/ServiceDiscovery+TypeErased.swift

```diff
@@ -176,6 +233,120 @@ public class AnyServiceDiscovery: ServiceDiscovery {
+
+    private let _subscribe: (AnyHashable, @escaping (Result<[AnyHashable], Error>) -> Void, @escaping (CompletionReason) -> Void) -> CancellationToken
+
+    public var defaultLookupTimeout: DispatchTimeInterval {
+        self._defaultLookupTimeout()
+    }
+
+    public init<ServiceDiscoveryImpl: AsyncServiceDiscovery>(_ serviceDiscovery: ServiceDiscoveryImpl) {
+        self._underlying = serviceDiscovery
+        self._defaultLookupTimeout = { serviceDiscovery.defaultLookupTimeout }
+
+        self._lookup = { anyService, deadline async throws in
+            guard let service = anyService.base as? ServiceDiscoveryImpl.Service else {
+                preconditionFailure("Expected service type to be \(ServiceDiscoveryImpl.Service.self), got \(type(of: anyService.base))")
+            }
+            let instances = try await serviceDiscovery.lookup(service, deadline: deadline)
+            return instances.map(AnyHashable.init)
+        }
+        self._subscribe = { anyService, nextResultHandler, completionHandler in
+            guard let service = anyService.base as? ServiceDiscoveryImpl.Service else {
+                preconditionFailure("Expected service type to be \(ServiceDiscoveryImpl.Service.self), got \(type(of: anyService.base))")
+            }
+            return serviceDiscovery.subscribe(
+                to: service,
+                onNext: { result in nextResultHandler(result.map { $0.map(AnyHashable.init) }) },
+                onComplete: completionHandler
+            )
+        }
+    }
+
+    /// See `AsyncServiceDiscovery.lookup`.
+    ///
+    /// - Warning: If `service` type does not match the underlying `AsyncServiceDiscovery`'s, it would result in a failure.
+    public func lookup(_ service: AnyHashable, deadline: DispatchTime? = nil) async throws -> [AnyHashable] {
+        try await self._lookup(service, deadline)
+    }
+
+    /// See `lookup`.
+    ///
+    /// - Warning: If `Service` or `Instance` type does not match the underlying `AsyncServiceDiscovery`'s associated types, it would result in a failure.
+    public func lookupAndUnwrap<Service, Instance>(_ service: Service, deadline: DispatchTime? = nil) async throws -> [Instance] where Service: Hashable, Instance: Hashable {
+        let anyInstances = try await self._lookup(AnyHashable(service), deadline)
+        return try self.transform(anyInstances)
+    }
+
+    /// See `ServiceDiscovery.subscribe`.
+    ///
+    /// - Warning: If `service` type does not match the underlying `AsyncServiceDiscovery`'s, it would result in a failure.
+    @discardableResult
+    public func subscribe(
+        to service: AnyHashable,
+        onNext nextResultHandler: @escaping (Result<[AnyHashable], Error>) -> Void,
+        onComplete completionHandler: @escaping (CompletionReason) -> Void = { _ in }
+    ) -> CancellationToken {
+        self._subscribe(service, nextResultHandler, completionHandler)
+    }
```

Comment:
This will eventually be able to use the `AsyncSequence` though to not be a huge pain to implement it might want https://github.com/apple/swift/pull/36921 which isn't through evolution yet so let's hold off on it... just FYI though.